c/frame/_repr_html_.cc

```diff
@@ -0,0 +1,270 @@
+//------------------------------------------------------------------------------
+// Copyright 2018 H2O.ai
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//------------------------------------------------------------------------------
+#include "frame/py_frame.h"
+#include <sstream>
+#include "python/string.h"
+#include "types.h"
+
+
+
+class HtmlWidget {
+  private:
+    std::ostringstream html;
+    DataTable* dt;
+    size_t ncols, cols0, cols1;
+    size_t nrows, rows0, rows1;
+    static bool styles_emitted;
+
+  public:
+    HtmlWidget(DataTable* dt_) {
+      const size_t maxcols = 30;  // TODO: make configurable
+      const size_t maxrows = 45;
+      dt = dt_;
+      ncols = static_cast<size_t>(dt->ncols);
+      nrows = static_cast<size_t>(dt->nrows);
+      cols0 = ncols <= maxcols ? ncols : maxcols * 2 / 3;
+      cols1 = ncols <= maxcols ? 0 : maxcols - cols0;
+      rows0 = nrows <= maxrows ? nrows : maxrows * 2 / 3;
+      rows1 = nrows <= maxrows ? 0 : maxrows - rows0;
+    }
+
+    py::oobj to_pystring() {
+      render_all();
+      const std::string htmlstr = html.str();
+      return py::ostring(htmlstr);
+    }
+
+
+  private:
+    void render_all() {
+      render_styles();
+      html << "<div class=datatable>";
```

Comment:
I would quote attributes even though this is not required. Several reasons: 
- looks cleaner as you can quickly distinguish where the value starts/ends;
- later you can always do `<div class='datatable something_else'>` without the need to add quotes, so no way to forget to do that.

It's a matter of taste, of course.